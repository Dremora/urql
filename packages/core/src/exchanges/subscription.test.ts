import { print } from 'graphql';
import {
  empty,
  publish,
  fromValue,
  pipe,
  Source,
  take,
  toPromise,
  scan,
} from 'wonka';
import { Client } from '../client';
import { gql } from '../gql';
import { makeOperation } from '../utils';
import {
  context,
  subscriptionOperation,
  subscriptionResult,
} from '../test-utils';
import { Operation, OperationResult } from '../types';
import { subscriptionExchange, SubscriptionForwarder } from './subscription';

it('should return response data from forwardSubscription observable', async () => {
  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = operation => {
    expect(operation.query).toBe(print(subscriptionOperation.query));
    expect(operation.variables).toBe(subscriptionOperation.variables);
    expect(operation.context).toEqual(subscriptionOperation.context);

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = await pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    take(1),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalled();
});

it('should tear down the operation if the source subscription ends', async () => {
  const reexecuteOperation = jest.fn();
  const unsubscribe = jest.fn();

  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: { reexecuteOperation: reexecuteOperation as any } as Client,
  };

  const forwardSubscription: SubscriptionForwarder = () => ({
    subscribe(observer) {
      observer.complete();
      return { unsubscribe };
    },
  });

  pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    publish
  );

  await Promise.resolve();

  expect(unsubscribe).not.toHaveBeenCalled();
  expect(reexecuteOperation).toHaveBeenCalled();
});

it('should support deferred operations coming in', async () => {
  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = () => {
    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next({
            hasNext: true,
            data: {
              author: {
                id: '1',
                name: 'Steve',
                __typename: 'Author',
                todos: [{ id: '1', text: 'stream', __typename: 'Todo' }],
              },
            },
          });

          observer.next({
            path: ['author', 'todos', 1],
            data: { id: '2', text: 'defer', __typename: 'Todo' },
            hasNext: false,
          });

          observer.complete();
        });

        return { unsubscribe };
      },
    };
  };

  const streamedQueryOperation: Operation = makeOperation(
    'query',
    {
      query: gql`
        query {
          author {
            id
            name
            todos @stream {
              id
              text
            }
          }
        }
      `,
      variables: {},
      key: 1,
    },
    context
  );

  const chunks: OperationResult[] = await pipe(
    fromValue(streamedQueryOperation),
    subscriptionExchange({ forwardSubscription, enableAllOperations: true })(
      exchangeArgs
    ),
    scan((prev: OperationResult[], item) => [...prev, item], []),
    toPromise
  );

  expect(chunks).toMatchSnapshot();
});

it('should support streamed operations coming in', async () => {
  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = () => {
    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next({
            hasNext: true,
            data: {
              author: {
                id: '1',
                __typename: 'Author',
              },
            },
          });

          observer.next({
            path: ['author'],
            data: { name: 'Steve' },
            hasNext: true,
          });
          observer.complete();
        });

        return { unsubscribe };
      },
    };
  };
  const AuthorFragment = gql`
    fragment authorFields on Author {
      name
    }
  `;

  const streamedQueryOperation: Operation = makeOperation(
    'query',
    {
      query: gql`
        query {
          author {
            id
            ...authorFields @defer
          }
        }

        ${AuthorFragment}
      `,
      variables: {},
      key: 1,
    },
    context
  );
  const chunks: OperationResult[] = await pipe(
    fromValue(streamedQueryOperation),
    subscriptionExchange({ forwardSubscription, enableAllOperations: true })(
      exchangeArgs
    ),
    scan((prev: OperationResult[], item) => [...prev, item], []),
    toPromise
  );

  expect(chunks).toMatchSnapshot();
});
