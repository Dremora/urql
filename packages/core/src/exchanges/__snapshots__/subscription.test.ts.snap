// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return response data from forwardSubscription observable 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 4,
    "kind": "subscription",
    "query": Object {
      "__key": 2088253569,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "subscribeToUser",
          },
          "operation": "subscription",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "user",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "user",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 92,
        "source": Object {
          "body": "# subscribeToUser
subscription subscribeToUser($user: String) { user(user: $user) { name } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "user": "colin",
    },
  },
}
`;

exports[`should support deferred operations coming in 1`] = `
Array [
  Object {
    "data": Object {
      "author": Object {
        "__typename": "Author",
        "id": "1",
        "name": "Steve",
        "todos": Array [
          Object {
            "__typename": "Todo",
            "id": "1",
            "text": "stream",
          },
        ],
      },
    },
    "error": undefined,
    "extensions": undefined,
    "hasNext": true,
    "operation": Object {
      "context": Object {
        "fetchOptions": Object {
          "method": "POST",
        },
        "requestPolicy": "cache-first",
        "url": "http://localhost:3000/graphql",
      },
      "key": 1,
      "kind": "query",
      "query": Object {
        "__key": 3321206837,
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "loc": undefined,
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "author",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "loc": undefined,
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [
                          Object {
                            "arguments": Array [],
                            "kind": "Directive",
                            "loc": undefined,
                            "name": Object {
                              "kind": "Name",
                              "loc": undefined,
                              "value": "stream",
                            },
                          },
                        ],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "todos",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "loc": undefined,
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": undefined,
                              "name": Object {
                                "kind": "Name",
                                "loc": undefined,
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": undefined,
                              "name": Object {
                                "kind": "Name",
                                "loc": undefined,
                                "value": "text",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 48,
          "source": Object {
            "body": "{ author { id name todos @stream { id text } } }",
            "locationOffset": Object {
              "column": 1,
              "line": 1,
            },
            "name": "gql",
          },
          "start": 0,
        },
      },
      "variables": Object {},
    },
  },
  Object {
    "data": Object {
      "author": Object {
        "__typename": "Author",
        "id": "1",
        "name": "Steve",
        "todos": Array [
          Object {
            "__typename": "Todo",
            "id": "1",
            "text": "stream",
          },
          Object {
            "__typename": "Todo",
            "id": "2",
            "text": "defer",
          },
        ],
      },
    },
    "error": undefined,
    "extensions": undefined,
    "hasNext": false,
    "operation": Object {
      "context": Object {
        "fetchOptions": Object {
          "method": "POST",
        },
        "requestPolicy": "cache-first",
        "url": "http://localhost:3000/graphql",
      },
      "key": 1,
      "kind": "query",
      "query": Object {
        "__key": 3321206837,
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "loc": undefined,
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "author",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "loc": undefined,
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [
                          Object {
                            "arguments": Array [],
                            "kind": "Directive",
                            "loc": undefined,
                            "name": Object {
                              "kind": "Name",
                              "loc": undefined,
                              "value": "stream",
                            },
                          },
                        ],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "todos",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "loc": undefined,
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": undefined,
                              "name": Object {
                                "kind": "Name",
                                "loc": undefined,
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "loc": undefined,
                              "name": Object {
                                "kind": "Name",
                                "loc": undefined,
                                "value": "text",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 48,
          "source": Object {
            "body": "{ author { id name todos @stream { id text } } }",
            "locationOffset": Object {
              "column": 1,
              "line": 1,
            },
            "name": "gql",
          },
          "start": 0,
        },
      },
      "variables": Object {},
    },
  },
]
`;

exports[`should support streamed operations coming in 1`] = `
Array [
  Object {
    "data": Object {
      "author": Object {
        "__typename": "Author",
        "id": "1",
      },
    },
    "error": undefined,
    "extensions": undefined,
    "hasNext": true,
    "operation": Object {
      "context": Object {
        "fetchOptions": Object {
          "method": "POST",
        },
        "requestPolicy": "cache-first",
        "url": "http://localhost:3000/graphql",
      },
      "key": 1,
      "kind": "query",
      "query": Object {
        "__key": 496779830,
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "loc": undefined,
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "author",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "loc": undefined,
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "directives": Array [
                          Object {
                            "arguments": Array [],
                            "kind": "Directive",
                            "loc": undefined,
                            "name": Object {
                              "kind": "Name",
                              "loc": undefined,
                              "value": "defer",
                            },
                          },
                        ],
                        "kind": "FragmentSpread",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "authorFields",
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
          Object {
            "directives": Array [],
            "kind": "FragmentDefinition",
            "loc": Object {
              "end": 40,
              "source": Object {
                "body": "fragment authorFields on Author { name }",
                "locationOffset": Object {
                  "column": 1,
                  "line": 1,
                },
                "name": "gql",
              },
              "start": 0,
            },
            "name": Object {
              "kind": "Name",
              "loc": undefined,
              "value": "authorFields",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "typeCondition": Object {
              "kind": "NamedType",
              "loc": undefined,
              "name": Object {
                "kind": "Name",
                "loc": undefined,
                "value": "Author",
              },
            },
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 81,
          "source": Object {
            "body": "{ author { id ...authorFields @defer } } fragment authorFields on Author { name }",
            "locationOffset": Object {
              "column": 1,
              "line": 1,
            },
            "name": "gql",
          },
          "start": 0,
        },
      },
      "variables": Object {},
    },
  },
  Object {
    "data": Object {
      "author": Object {
        "__typename": "Author",
        "id": "1",
        "name": "Steve",
      },
    },
    "error": undefined,
    "extensions": undefined,
    "hasNext": true,
    "operation": Object {
      "context": Object {
        "fetchOptions": Object {
          "method": "POST",
        },
        "requestPolicy": "cache-first",
        "url": "http://localhost:3000/graphql",
      },
      "key": 1,
      "kind": "query",
      "query": Object {
        "__key": 496779830,
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "loc": undefined,
            "name": undefined,
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "author",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "loc": undefined,
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "directives": Array [
                          Object {
                            "arguments": Array [],
                            "kind": "Directive",
                            "loc": undefined,
                            "name": Object {
                              "kind": "Name",
                              "loc": undefined,
                              "value": "defer",
                            },
                          },
                        ],
                        "kind": "FragmentSpread",
                        "loc": undefined,
                        "name": Object {
                          "kind": "Name",
                          "loc": undefined,
                          "value": "authorFields",
                        },
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [],
          },
          Object {
            "directives": Array [],
            "kind": "FragmentDefinition",
            "loc": Object {
              "end": 40,
              "source": Object {
                "body": "fragment authorFields on Author { name }",
                "locationOffset": Object {
                  "column": 1,
                  "line": 1,
                },
                "name": "gql",
              },
              "start": 0,
            },
            "name": Object {
              "kind": "Name",
              "loc": undefined,
              "value": "authorFields",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": undefined,
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": undefined,
                  "name": Object {
                    "kind": "Name",
                    "loc": undefined,
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "typeCondition": Object {
              "kind": "NamedType",
              "loc": undefined,
              "name": Object {
                "kind": "Name",
                "loc": undefined,
                "value": "Author",
              },
            },
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 81,
          "source": Object {
            "body": "{ author { id ...authorFields @defer } } fragment authorFields on Author { name }",
            "locationOffset": Object {
              "column": 1,
              "line": 1,
            },
            "name": "gql",
          },
          "start": 0,
        },
      },
      "variables": Object {},
    },
  },
]
`;
